### PR TITLE
fix(app, robot-server): support `retryLocation` when retrying `dropTipInPlace` during Error Recovery

### PR DESCRIPTION
**Pull request by mjhuff**
Closes [RQA-3591](https://opentrons.atlassian.net/browse/RQA-3591)

# Overview

When dropping a tip in a trash bin or waste chute, the pipette moves downwards before doing the drop tip. During Error Recovery, when we retry the failed command with a `fixit` intent, we need to perform a `moveToCoordinates` first before dispatching the `dropTipInPlace` or the robot drops the tip(s) from too high a location. 

To know how far to move, we need the `failedCommand` to contain `retryLocation` metadata. We have a pattern for this already with `overpressure` in place commands, so we extend the same functionality here.

## Test Plan and Hands on Testing

* [x] Test with one of the protocols in RQA-3591.

## Changelog
- Fixed the "Retry tip drop" recovery option in "Tip drop failed" recovery flows dropping tips too high when retried over a waste chute or trash bin.

## Risk assessment
low

[RQA-3591]: https://opentrons.atlassian.net/browse/RQA-3591?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: api/src/opentrons/protocol_engine/commands/aspirate_in_place.py [hunk 1]
```diff
@@ -83,22 +83,22 @@ async def execute(self, params: AspirateInPlaceParams) -> _ExecuteReturn:
...
         if not ready_to_aspirate:
             raise PipetteNotReadyToAspirateError(
                 "Pipette cannot aspirate in place because of a previous blow out."
                 " The first aspirate following a blow-out must be from a specific well"
                 " so the plunger can be reset in a known safe position."
             )
 
-        state_update = StateUpdate()
         current_location = self._state_view.pipettes.get_current_location()
+        current_position = await self._gantry_mover.get_position(params.pipetteId)
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/aspirate_in_place.py:99]: Unrelated fix. Fixing a potentially unbound reference to `curent_position`.

File: api/src/opentrons/protocol_engine/commands/dispense_in_place.py [hunk 2]
```diff
@@ -76,8 +76,8 @@ async def execute(self, params: DispenseInPlaceParams) -> _ExecuteReturn:
         """Dispense without moving the pipette."""
         state_update = StateUpdate()
         current_location = self._state_view.pipettes.get_current_location()
+        current_position = await self._gantry_mover.get_position(params.pipetteId)
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/dispense_in_place.py:79]: Unrelated fix. Fixing a potentially unbound reference to `current_position`.

File: api/src/opentrons/protocol_engine/commands/drop_tip.py [hunk 3]
```diff
@@ -145,6 +145,7 @@ async def execute(self, params: DropTipParams) -> _ExecuteReturn:
                         error=exception,
                     )
                 ],
+                errorInfo={"retryLocation": position},
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/drop_tip.py:148]: This is redundant for `dropTip`, but it lets us share the same `TipPhysicallyAttachedError` shape.

File: api/src/opentrons/protocol_engine/commands/drop_tip_in_place.py [hunk 4]
```diff
@@ -57,15 +57,19 @@ def __init__(
...
     ) -> None:
         self._tip_handler = tip_handler
         self._model_utils = model_utils
+        self._gantry_mover = gantry_mover
 
     async def execute(self, params: DropTipInPlaceParams) -> _ExecuteReturn:
         """Drop a tip using the requested pipette."""
         state_update = update_types.StateUpdate()
 
+        retry_location = await self._gantry_mover.get_position(params.pipetteId)
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/drop_tip_in_place.py:71]: @mjhuff Do we ever attempt a `dropTipInPlace` in a situation where the robot is not homed? If so, this might cause trouble, because the robot doesn't know the `retryLocation`, and my naive implementation here will cause it to raise an exception and fail the command. I think we had another bug like this.

**Review by SyntaxColoring — APPROVED** [review]
Frontend changes LGTM, thanks.

I wrote the backend changes, so I can't approve those, and also see my question above.

File: api/src/opentrons/protocol_engine/commands/drop_tip_in_place.py (hunk 4 above)
Comment (mjhuff) [reply api/src/opentrons/protocol_engine/commands/drop_tip_in_place.py:71]: > @mjhuff Do we ever attempt a `dropTipInPlace` in a situation where the robot is not homed?

Hm, it's possible for users to `dropTipInPlace` during the drop tip wizard portion of recovery (so effectively every flow minus the gripper flow) and during the "Tip drop failed" flow. 

However, in theory, the only instances during ER that the robot may be unaware of its position is during a gripper stall/collision error, and it's possible to get to dt wiz during that flow. 

We do update position estimators before the blowout/drop tip, however, if a `MustHomeError` is possible, this will be problematic, since error recovery has its own error modals and won't show the "home the gantry" CTA.

Comment (sfoster1) [reply api/src/opentrons/protocol_engine/commands/drop_tip_in_place.py:71]: If you tried to `dropTipInPlace` when you don't have a position estimation then it will fail, so if we're ever doing that then it isn't working already. After you `unsafe/updatePositionEstimators`, `get_position` will work.

**Review by sfoster1 — APPROVED** [review]
Looks good to me!

File: api/src/opentrons/protocol_engine/commands/drop_tip_in_place.py (hunk 4 above)
Comment (mjhuff) [reply api/src/opentrons/protocol_engine/commands/drop_tip_in_place.py:71]: Ok, great, I believe we are covered then. If you're dropping tips during a gripper stall/collision, you are about to end the run, and `get_position` won't matter. If you are resuming the run after a gripper stall/collision, we are doing an "everything but the plungers" home. 

If you are `dropTipInPlace`ing during a different error recovery flow, as of this release, the robot knows its position.

**Comment by SyntaxColoring** [comment]
On hold until we fix https://opentrons.atlassian.net/browse/RQA-3599 and we can test this with gripper error recovery.

**Comment by mjhuff** [comment]
> On hold until we fix https://opentrons.atlassian.net/browse/RQA-3599 and we can test this with gripper error recovery.

Done!
